lightning/src/util/ser_macros.rs

```diff
@@ -144,6 +147,9 @@ macro_rules! check_missing_tlv {
 	($last_seen_type: expr, $type: expr, $field: ident, ignorable) => {{
 		// no-op
 	}};
+	($last_seen_type: expr, $type: expr, $field: ident, (read_arg, $read_arg: expr)) => {{
```

Comment:
Fine with that. @jkczyz thoughts?

Reply:
Hmm... maybe we can make this less of a special case and instead as a variation of `option` that gives the trait as an `ident` macro param. Something like:

```rust
decode_tlv_stream!(&writer.0[..], {
    (1, read_adapter, option: (LengthReadableArgs, arg: rho)),
});
```

Syntax-wise, I think I suggested awhile ago we changed `(default_value, $default: expr)` to `(default: $default: expr)`, which this would align better with. I guess the exact syntax isn't too important now.

Reply:
Maybe even drop the `arg: ` token such that non-`Args` traits could be used without special casing. Don't have a strong opinion on that part.

Reply:
I had to move the `(` outside of the option to work with the `tt` type in `decode_tlv_stream`, but this seems to work.

```diff
diff --git a/lightning/src/util/chacha20poly1305rfc.rs b/lightning/src/util/chacha20poly1305rfc.rs
index 3b2d2627..5a8727cc 100644
--- a/lightning/src/util/chacha20poly1305rfc.rs
+++ b/lightning/src/util/chacha20poly1305rfc.rs
@@ -393,7 +393,7 @@ mod tests {
                // Now deserialize the object back and make sure it matches the original.
                let mut read_adapter: Option<ChaChaPolyReadAdapter<TestWriteable>> = None;
                decode_tlv_stream!(&writer.0[..], {
-                       (1, read_adapter, (read_arg, rho)),
+                       (1, read_adapter, (option: LengthReadableArgs, rho)),
                });
                assert_eq!(writeable, read_adapter.unwrap().readable);
 
diff --git a/lightning/src/util/ser_macros.rs b/lightning/src/util/ser_macros.rs
index 35e4ecf6..816426b1 100644
--- a/lightning/src/util/ser_macros.rs
+++ b/lightning/src/util/ser_macros.rs
@@ -118,7 +118,7 @@ macro_rules! check_tlv_order {
        ($last_seen_type: expr, $typ: expr, $type: expr, $field: ident, ignorable) => {{
                // no-op
        }};
-       ($last_seen_type: expr, $typ: expr, $type: expr, $field: ident, (read_arg, $read_arg: expr)) => {{
+       ($last_seen_type: expr, $typ: expr, $type: expr, $field: ident, (option: $trait: ident $(, $read_arg: expr)?)) => {{
                // no-op
        }};
 }
@@ -147,7 +147,7 @@ macro_rules! check_missing_tlv {
        ($last_seen_type: expr, $type: expr, $field: ident, ignorable) => {{
                // no-op
        }};
-       ($last_seen_type: expr, $type: expr, $field: ident, (read_arg, $read_arg: expr)) => {{
+       ($last_seen_type: expr, $type: expr, $field: ident, (option: $trait: ident $(, $read_arg: expr)?)) => {{
                // no-op
        }};
 }
@@ -169,9 +169,8 @@ macro_rules! decode_tlv {
        ($reader: expr, $field: ident, ignorable) => {{
                $field = ser::MaybeReadable::read(&mut $reader)?;
        }};
-       // Read a ser::LengthReadableArgs
-       ($reader: expr, $field: ident, (read_arg, $read_arg: expr)) => {{
-               $field = Some(ser::LengthReadableArgs::read(&mut $reader, $read_arg)?);
+       ($reader: expr, $field: ident, (option: $trait: ident $(, $read_arg: expr)?)) => {{
+               $field = Some($trait::read(&mut $reader $(, $read_arg)*)?);
        }};
 }
```

Reply:
Thanks Jeff, added that! Sorry, I forgot to add a fixup commit :(